shell/browser/extensions/electron_extension_loader.cc

```diff
@@ -155,9 +156,10 @@ void ElectronExtensionLoader::LoadExtensionForReload(
     LoadErrorBehavior load_error_behavior) {
   CHECK(!path.empty());
 
+  int load_flags = Extension::FOLLOW_SYMLINKS_ANYWHERE; // FIXME
```

Comment:
I think this is a place where `ExtensionPrefs` would be useful. Looking at [Chrome's installed extension loader](https://source.chromium.org/chromium/chromium/src/+/master:chrome/browser/extensions/installed_loader.cc;l=687;drc=02fc05043bd38d318ba9ce59ccc4c7602284f990), they use this to store global and specific extension preferences at runtime—including [methods for getting/setting file access.](https://source.chromium.org/chromium/chromium/src/+/master:extensions/browser/extension_prefs.h;l=517;drc=3c2bf33cc5112e89df2ebc668def33b9363f1723)

In `Session::LoadExtension`, we could call `ExtensionPrefs::SetAllowFileAccess` with the value of the `allowFileAccess` option. Then `ElectronExtensionLoader` can derive the `load_flags` from `ExtensionPrefs`. Chrome's [`InstalledLoader::GetCreationFlags`](https://source.chromium.org/chromium/chromium/src/+/master:chrome/browser/extensions/installed_loader.cc;l=687;drc=02fc05043bd38d318ba9ce59ccc4c7602284f990) is a good example of how this could work.

Reply:
I thought about that, but the thing I noticed while testing (https://github.com/electron/electron/issues/24011#issuecomment-674315764) is that:
1) This preference doesn't appear to do anything on it's own, or it could do something I don't know yet.
2) In Chromium, this is apparently not what is used.
  * A call to `ExtensionPrefs::SetAllowFileAccess` writes to `{profile}/Preferences`.
  * But in Chromium that doesn't happen when file access is enabled in extension preferences.
  * Instead, Chromium stores that in`{profile}/Current Session` which is inexistent in Electron.

My concern here is that perhaps `ExtensionPrefs::SetAllowFileAccess` does something _else_ and can have some other side effect? Or is that not the case?

I can surely port to that if it would be safe to use it here.

Reply:
After looking more into this, Chrome implements a private API on the chrome://extensions page with a function `chrome.developerPrivate.updateExtensionConfiguration(config, callback)`. [It's internal implementation](https://source.chromium.org/chromium/chromium/src/+/master:chrome/browser/extensions/api/developer_private/developer_private_api.cc;l=891;drc=f8990f540193a784158e5e4d9ec3ac6d1cd76771) calls a utility which in turn calls `ExtensionPrefs::SetAllowFileAccess`.

After this function is called, the value will stay in memory and (I think) persist to disk soon after. An extension loader implements a [`GetCreationFlags`](https://source.chromium.org/chromium/chromium/src/+/master:chrome/browser/extensions/installed_loader.cc;l=687;drc=02fc05043bd38d318ba9ce59ccc4c7602284f990) method to then derive flags from `ExtensionPrefs`.

In my opinion, sticking with this pattern close to Chrome will allow us to more naturally build on the extension support in Electron later on.

As for the the location of prefs (`{profile}/Preferences`), I'm not sure what the right path would be us. I think we can go with the default and change it later if needed.

Reply:
I don't believe Electron persists such preferences to disk. Additionally, I'd prefer that the extensions loading stuff doesn't persist anything to disk internally, as that's often surprising for developers (e.g. "i removed the `loadExtension(...)` line from my app, why is the react devtools extension still loading??" behavior that used to be present.)

Reply:
Update on `ExtensionPrefs`, they're mostly used by Chrome to persist extension settings for a profile.
- Whether an extension has file access or can run in incognito mode.
- The extension installation time.
- If an extension is disabled, the reason why.
- And many other Chrome-specific settings.

These, among many other preferences, aren't too relevant to an Electron app and could be managed outside of Electron internal code. However, special cases such as [`chrome.runtime.reload()`](https://developer.chrome.com/extensions/runtime#method-reload) will cause this setting to be dropped without relying on storage such as `ExtensionPrefs`.

Additionally, `ExtensionPrefs` _do_ end up persisting to disk using the same `JsonPrefStore` as `UserPrefs` which are setup in [`ElectronBrowserContext::InitPrefs`](https://github.com/electron/electron/blob/master/shell/browser/electron_browser_context.cc#L169-L185).

To avoid this behavior, I believe it will be necessary to create an [`InMemoryPrefStore`](https://source.chromium.org/chromium/chromium/src/+/master:components/prefs/in_memory_pref_store.h) exclusively for ExtensionPrefs which would be returned from `ElectronExtensionsBrowserClient::GetPrefServiceForContext`. Possibly a command-line flag could be used if a developer wanted to opt-in to persisting these settings?

Reply:
@ChALkeR to followup, we can simplify this by avoiding `ExtensionPrefs` altogether. Adding a map of `extensionId -> allowFileAccess` could be added to `ElectronExtensionLoader`. When an extension is unloaded, we can clear it. That way we can allow persisting file access if an extension is reloaded.